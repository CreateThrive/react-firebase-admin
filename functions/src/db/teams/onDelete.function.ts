import { database } from 'firebase-admin';
import { database as db } from 'firebase-functions';
import { BatchUpdate, IUserRtdb } from '../../types';

export default db.ref('teams/{teamId}').onDelete(async (snapshot, context) => {
  const { teamId } = context.params;
  let users;

  try {
    users = (await database().ref('users').once('value')).val();
  } catch (error) {
    console.log(error);
  }

  users = Object.entries(users).filter(([key, user]) => {
    const { teams } = user as IUserRtdb;

    return teams?.includes(teamId);
  });

  const batchUpdate: BatchUpdate = {};

  users.forEach(([key, user]) => {
    const { teams } = user as IUserRtdb;

    if (teams) {
      const newTeams = teams.filter((index: string) => index !== teamId);

      batchUpdate[`/${key}/teams`] = newTeams;
    }
  });

  return database().ref('users').update(batchUpdate);
});
