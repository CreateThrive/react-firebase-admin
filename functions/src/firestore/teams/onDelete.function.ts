import { firestore } from 'firebase-admin';
import { firestore as firestoreFunction } from 'firebase-functions';
import { IUserFirestore } from '../../types';

export default firestoreFunction
  .document('teams/{teamId}')
  .onDelete(async (snapshot, context) => {
    const { teamId } = context.params;

    const batch = firestore().batch();

    let snapshotsUsers;
    try {
      snapshotsUsers = await firestore()
        .collection('users')
        .where('teams', 'array-contains', teamId)
        .get();
    } catch (error) {
      console.log(error);
    }

    snapshotsUsers?.forEach((userSnapshot: any) => {
      const { id } = userSnapshot;
      const { teams } = userSnapshot.data() as IUserFirestore;

      const newTeams = teams.filter((index: string) => index !== teamId);

      const userRef = firestore().collection('users').doc(id);

      batch.update(userRef, { teams: newTeams });
    });

    return batch.commit();
  });
