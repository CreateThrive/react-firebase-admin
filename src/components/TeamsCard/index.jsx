/* eslint-disable no-nested-ternary */
/* eslint-disable react/forbid-prop-types */
import React, { useState, useEffect } from 'react';
import { useDispatch, useSelector, shallowEqual } from 'react-redux';
import ClipLoader from 'react-spinners/ClipLoader';

import { useFormatMessage } from 'hooks';
import MediaCard from 'components/MediaCard';
import ConfirmationModal from 'components/ConfirmationModal';
import TeamModal from 'components/TeamModal';
import {
  deleteTeam,
  fetchTeams,
  fetchTeamUsers,
  teamsCleanUp,
} from 'state/actions/teams';
import { closeModal, openModal } from 'utils';

const TeamsCard = () => {
  const { success, loading, isAdmin, deleted, dataList } = useSelector(
    (state) => ({
      isAdmin: state.auth.userData.isAdmin,
      success: state.teams.success,
      loading: state.teams.loadingTeams,
      deleted: state.teams.deleted,
      dataList: state.teams.teamsList,
    }),
    shallowEqual
  );

  const dispatch = useDispatch();

  const [search, setSearch] = useState('');

  useEffect(() => {
    dispatch(fetchTeams());

    return () => {
      dispatch(teamsCleanUp());
    };
  }, [dispatch]);

  const data = search
    ? dataList.filter((el) => {
        const clonedElem = { ...el };
        delete clonedElem.id;
        delete clonedElem.isAdmin;
        delete clonedElem.logoUrl;
        delete clonedElem.teams;
        delete clonedElem.users;
        delete clonedElem.description;
        delete clonedElem.createdAt;
        delete clonedElem.createdBy;
        return Object.values(clonedElem).some((field) =>
          field.toLowerCase().includes(search.toLowerCase())
        );
      })
    : dataList;

  const availableTeamsMessage = useFormatMessage('TeamsCard.allTeams');

  const addTeamMessage = useFormatMessage('TeamsCard.addTeam');

  const searchByMessage = useFormatMessage('TeamsCard.searchBy');

  const confirmMessage = useFormatMessage('TeamsCard.confirm');

  const deleteMessage = useFormatMessage('TeamsCard.delete');

  const permDeleteTeamMessage = useFormatMessage('TeamsCard.permDeleteTeam');

  const cancelMessage = useFormatMessage('TeamsCard.cancel');

  const [deleteModal, setDeleteModal] = useState({
    id: null,
    isOpen: false,
  });

  const [addModal, setAddModal] = useState({
    id: null,
    isOpen: false,
  });

  const [isEditing, setIsEditing] = useState(false);

  const onCloseDeleteModalHandler = () => {
    closeModal(setDeleteModal);
    setIsEditing(false);
  };

  const onCloseAddModalHandler = () => {
    closeModal(setAddModal);
    setIsEditing(false);
  };

  useEffect(() => {
    if (deleted && !loading) {
      onCloseDeleteModalHandler();
    }
  }, [deleted, loading]);

  useEffect(() => {
    if (success && !loading) {
      onCloseAddModalHandler();
    }
  }, [success, loading]);

  const onRemoveButtonClickHandler = (id) => {
    openModal(setDeleteModal, id);
  };

  const onAddButtonClickHandler = (id) => {
    openModal(setAddModal, id);
  };

  const onModifyButtonClickHandler = (id) => {
    openModal(setAddModal, id);
    setIsEditing(true);
  };

  const onDeleteHandler = () => {
    const { id } = deleteModal;
    dispatch(deleteTeam(id));
    const team = dataList.find((value) => value.id !== id);
    if (team) {
      dispatch(fetchTeamUsers(team.id));
    }
  };

  return (
    <>
      {deleteModal.isOpen && (
        <ConfirmationModal
          isActive={deleteModal.isOpen}
          isLoading={loading}
          confirmButtonMessage={deleteMessage}
          title={confirmMessage}
          body={permDeleteTeamMessage}
          cancelButtonMessage={cancelMessage}
          onConfirmation={onDeleteHandler}
          onCancel={onCloseDeleteModalHandler}
        />
      )}
      {addModal.isOpen && (
        <TeamModal
          isActive={addModal.isOpen}
          onCancel={onCloseAddModalHandler}
          isEditing={isEditing}
        />
      )}
      <div className="card has-height-medium">
        <header className="card-header">
          <p className="card-header-title">
            <span className="icon">
              <i className="mdi mdi-account-group default" />
            </span>
            <span>{availableTeamsMessage}</span>
          </p>
          {isAdmin && (
            <button
              type="button"
              className="button"
              onClick={onAddButtonClickHandler}
            >
              <span className="icon">
                <i className="mdi mdi-account-multiple-plus default" />
              </span>
              <span>{addTeamMessage}</span>
            </button>
          )}
        </header>
        <div>
          <div
            className="field-body"
            style={{
              padding: '0.5rem 0.25rem',
              borderBottom: '1px solid rgba(24,28,33,0.06)',
            }}
          >
            <div className="field">
              <div className="control is-expanded">
                <input
                  placeholder={searchByMessage}
                  type="text"
                  className="input search"
                  value={search}
                  onChange={(e) => setSearch(e.target.value)}
                  style={{ height: '38px' }}
                />
              </div>
            </div>
            <div className="field">
              <div className="control is-expanded">
                <div style={{ minWidth: '220px' }} />
              </div>
            </div>
          </div>
        </div>
        <div
          className="card-content ps ps--active-y"
          style={{
            padding: '0px 1.5rem',
            overflow: 'auto',
            maxHeight: '700px',
          }}
        >
          {loading ? (
            <ClipLoader />
          ) : (
            <div className="media-list">
              {data &&
                data.map((value) => (
                  <MediaCard
                    content={value}
                    key={value.id}
                    onDeleteHandler={onRemoveButtonClickHandler}
                    onModifyHandler={onModifyButtonClickHandler}
                  />
                ))}
            </div>
          )}
        </div>
      </div>
    </>
  );
};

export default TeamsCard;
