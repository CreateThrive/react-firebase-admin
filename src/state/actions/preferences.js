import { createAction } from 'redux-act';

export const PREFERENCES_SET_LOCALE = createAction('PREFERENCES_SET_LOCALE');

export const setUserLocale = locale => dispatch => {
  return dispatch(PREFERENCES_SET_LOCALE({ locale }));
};
