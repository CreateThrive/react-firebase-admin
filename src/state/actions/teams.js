import { createAction } from 'redux-act';
import { toastr } from 'react-redux-toastr';
import { v4 as uuid } from 'uuid';

import { firebaseError, deleteLogo, getLogoUrl, uploadLogo } from 'utils';
import {
  fetchCollection,
  createDocument,
  deleteDocument,
  updateDocument,
  batchUpdateDocument,
} from '../api';
import { removeStoreUser, addStoreUser } from './users';

export const TEAMS_FETCH_TEAMS_INIT = createAction('TEAMS_FETCH_TEAMS_INIT');

export const TEAMS_FETCH_TEAMS_SUCCESS = createAction(
  'TEAMS_FETCH_TEAMS_SUCCESS'
);

export const TEAMS_FETCH_TEAMS_FAIL = createAction('TEAMS_FETCH_TEAMS_FAIL');

export const TEAMS_DELETE_TEAM_INIT = createAction('TEAMS_DELETE_TEAM_INIT');

export const TEAMS_DELETE_TEAM_SUCCESS = createAction(
  'TEAMS_DELETE_TEAM_SUCCESS'
);

export const TEAMS_DELETE_TEAM_FAIL = createAction('TEAMS_DELETE_TEAM_FAIL');

export const TEAMS_CREATE_TEAM_INIT = createAction('TEAMS_CREATE_TEAM_INIT');

export const TEAMS_CREATE_TEAM_SUCCESS = createAction(
  'TEAMS_CREATE_TEAM_SUCCESS'
);
export const TEAMS_CREATE_TEAM_FAIL = createAction('TEAMS_CREATE_TEAM_FAIL');

export const TEAMS_MODIFY_TEAM_INIT = createAction('TEAMS_MODIFY_TEAM_INIT');

export const TEAMS_MODIFY_TEAM_SUCCESS = createAction(
  'TEAMS_MODIFY_TEAM_SUCCESS'
);
export const TEAMS_MODIFY_TEAM_FAIL = createAction('TEAMS_MODIFY_TEAM_FAIL');

export const TEAMS_CLEAN_UP = createAction('TEAMS_CLEAN_UP');

export const TEAMS_ADD_USER_INIT = createAction('TEAMS_ADD_USER_INIT');

export const TEAMS_ADD_USER_SUCCESS = createAction('TEAMS_ADD_USER_SUCCESS');

export const TEAMS_ADD_USER_FAIL = createAction('TEAMS_ADD_USER_FAIL');

export const TEAMS_REMOVE_USER_INIT = createAction('TEAMS_REMOVE_USER_INIT');

export const TEAMS_REMOVE_USER_SUCCESS = createAction(
  'TEAMS_REMOVE_USER_SUCCESS'
);

export const TEAMS_REMOVE_USER_FAIL = createAction('TEAMS_REMOVE_USER_FAIL');

export const TEAMS_SELECTED_TEAM = createAction('TEAMS_SELECTED_TEAM');

export const TEAMS_FETCH_TEAM_USERS_INIT = createAction(
  'TEAMS_FETCH_TEAM_USERS_INIT'
);

export const TEAMS_FETCH_TEAM_USERS_SUCCESS = createAction(
  'TEAMS_FETCH_TEAM_USERS_SUCCESS'
);

export const TEAMS_FETCH_TEAM_USERS_FAIL = createAction(
  'TEAMS_FETCH_TEAM_USERS_FAIL'
);

export const fetchTeams = () => {
  return async (dispatch, getState) => {
    dispatch(TEAMS_FETCH_TEAMS_INIT());
    const { locale } = getState().preferences;
    const { isAdmin, id } = getState().auth.userData;

    let teams;
    try {
      teams = await fetchCollection(
        'teams',
        isAdmin
          ? {
              filterBy: 'users',
              value: id,
              isArray: true,
            }
          : {}
      );
    } catch (error) {
      const errorMessage = firebaseError(error.code, locale);
      toastr.error('', errorMessage);
      return dispatch(TEAMS_FETCH_TEAMS_FAIL({ error }));
    }

    return dispatch(
      TEAMS_FETCH_TEAMS_SUCCESS({
        teams: isAdmin
          ? teams
          : teams.filter((value) => value.users?.includes(id)),
      })
    );
  };
};

export const deleteTeam = (id) => {
  return async (dispatch, getState) => {
    dispatch(TEAMS_DELETE_TEAM_INIT());
    const { locale } = getState().preferences;
    const { logoUrl } = getState().teams.teamsList.find(
      (team) => team.id === id
    );

    const deleteLogoTask = logoUrl ? deleteLogo(logoUrl, 'teams') : null;

    const deleteTeamTask = deleteDocument('teams', id);

    try {
      await Promise.all([deleteLogoTask, deleteTeamTask]);
    } catch (error) {
      const errorMessage = firebaseError(error.code, locale);
      toastr.error('', errorMessage);
      return dispatch(
        TEAMS_DELETE_TEAM_FAIL({
          error: errorMessage,
        })
      );
    }

    toastr.success('', 'The team was deleted.');
    return dispatch(TEAMS_DELETE_TEAM_SUCCESS({ id }));
  };
};

export const createTeam = ({ name, description, file }) => {
  return async (dispatch, getState) => {
    dispatch(TEAMS_CREATE_TEAM_INIT());
    const { locale } = getState().preferences;

    const uid = uuid();

    let uploadLogoTask = null;
    let logoUrl = null;
    if (file) {
      logoUrl = getLogoUrl(uid, file, 'teams');
      uploadLogoTask = uploadLogo(uid, file, 'teams');
    }

    const teamData = {
      name,
      description,
      logoUrl,
      createdAt: new Date().toString(),
      createdBy: getState().auth.userData.name,
    };

    const createTeamDbTask = createDocument('teams', uid, teamData);

    try {
      await Promise.all([uploadLogoTask, createTeamDbTask]);
    } catch (error) {
      const errorMessage = firebaseError(error.code, locale);
      toastr.error('', errorMessage);
      return dispatch(
        TEAMS_CREATE_TEAM_FAIL({
          error: errorMessage,
        })
      );
    }

    toastr.success('', 'Team created successfully');
    return dispatch(
      TEAMS_CREATE_TEAM_SUCCESS({ team: { ...teamData, id: uid } })
    );
  };
};

export const modifyTeam = ({ name, description, file, id }) => {
  return async (dispatch, getState) => {
    dispatch(TEAMS_MODIFY_TEAM_INIT());
    const { locale } = getState().preferences;
    const { logoUrl } = getState().teams.teamsList.find(
      (team) => team.id === id
    );

    let deleteLogoTask;
    let uploadLogoTask;
    let newLogoUrl = null;
    if (file) {
      newLogoUrl = getLogoUrl(id, file, 'teams');
      deleteLogoTask = logoUrl && deleteLogo(logoUrl, 'teams');
      uploadLogoTask = uploadLogo(id, file, 'teams');
    }

    const teamData = {
      name,
      description,
      logoUrl: logoUrl || newLogoUrl,
    };
    const updateTeamDbTask = updateDocument('teams', id, teamData);

    try {
      await Promise.all([deleteLogoTask, uploadLogoTask, updateTeamDbTask]);
    } catch (error) {
      const errorMessage = firebaseError(error.code, locale);
      toastr.error('', errorMessage);
      return dispatch(
        TEAMS_MODIFY_TEAM_FAIL({
          error: errorMessage,
        })
      );
    }

    toastr.success('', 'Team updated successfully');

    return dispatch(TEAMS_MODIFY_TEAM_SUCCESS({ team: teamData, id }));
  };
};

export const teamsCleanUp = () => (dispatch) => dispatch(TEAMS_CLEAN_UP());

export const addUser = (teamId, user) => {
  return async (dispatch, getState) => {
    dispatch(TEAMS_ADD_USER_INIT());
    const { locale } = getState().preferences;
    const { teamsList } = getState().teams;

    const userId = user.id;

    const teamUsers =
      teamsList.find((value) => value.id === teamId)?.users || [];
    const userTeams = user?.teams || [];

    const newUserTeams = [...userTeams, teamId];
    const newTeamUsers = [...teamUsers, userId];

    const batchCollections = {
      users: {
        documents: [
          {
            id: userId,
            field: 'teams',
            value: newUserTeams,
          },
        ],
      },

      teams: {
        documents: [
          {
            id: teamId,
            field: 'users',
            value: newTeamUsers,
          },
        ],
      },
    };

    try {
      await batchUpdateDocument(batchCollections);
    } catch (error) {
      const errorMessage = firebaseError(error.code, locale);
      toastr.error('', errorMessage);
      return dispatch(TEAMS_ADD_USER_FAIL({ error }));
    }

    dispatch(removeStoreUser(userId));

    return dispatch(
      TEAMS_ADD_USER_SUCCESS({
        teamId,
        users: newTeamUsers,
        user: { ...user, teams: newUserTeams },
      })
    );
  };
};

export const removeUser = (teamId, userId) => {
  return async (dispatch, getState) => {
    dispatch(TEAMS_REMOVE_USER_INIT());
    const { locale } = getState().preferences;
    const { teamsList, usersList } = getState().teams;

    const user = usersList.find((value) => value.id === userId);
    const teamUsers =
      teamsList.find((value) => value.id === teamId)?.users || [];
    const userTeams = user?.teams || [];

    const newUserTeams = userTeams.filter((value) => value !== teamId);
    const newTeamUsers = teamUsers.filter((value) => value !== userId);

    const bulkUpdate = {};
    bulkUpdate[`/users/${userId}/teams`] = newUserTeams;
    bulkUpdate[`/teams/${teamId}/users`] = newTeamUsers;

    try {
      await updateDocument('', '', bulkUpdate);
    } catch (error) {
      const errorMessage = firebaseError(error.code, locale);
      toastr.error('', errorMessage);
      return dispatch(TEAMS_REMOVE_USER_FAIL({ error }));
    }

    dispatch(addStoreUser({ ...user, teams: newUserTeams }));

    return dispatch(
      TEAMS_REMOVE_USER_SUCCESS({
        teamId,
        users: newTeamUsers,
        userId,
      })
    );
  };
};

export const fetchTeamUsers = (teamId) => {
  return async (dispatch, getState) => {
    dispatch(TEAMS_FETCH_TEAM_USERS_INIT());
    const { locale } = getState().preferences;

    let users;
    try {
      users = await fetchCollection('users', {
        filterBy: 'users',
        value: teamId,
        isArray: true,
      });
    } catch (error) {
      const errorMessage = firebaseError(error.code, locale);
      toastr.error('', errorMessage);
      return dispatch(TEAMS_FETCH_TEAM_USERS_FAIL({ error }));
    }

    dispatch(TEAMS_SELECTED_TEAM({ teamId }));

    return dispatch(
      TEAMS_FETCH_TEAM_USERS_SUCCESS({
        users: users.filter((value) => value.teams?.includes(teamId)),
      })
    );
  };
};
