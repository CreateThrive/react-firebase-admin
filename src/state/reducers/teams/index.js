import { createReducer } from 'redux-act';

import {
  TEAMS_FETCH_TEAMS_INIT,
  TEAMS_FETCH_TEAMS_SUCCESS,
  TEAMS_FETCH_TEAMS_FAIL,
  TEAMS_DELETE_TEAM_INIT,
  TEAMS_DELETE_TEAM_SUCCESS,
  TEAMS_DELETE_TEAM_FAIL,
  TEAMS_CREATE_TEAM_INIT,
  TEAMS_CREATE_TEAM_SUCCESS,
  TEAMS_CREATE_TEAM_FAIL,
  TEAMS_MODIFY_TEAM_INIT,
  TEAMS_MODIFY_TEAM_SUCCESS,
  TEAMS_MODIFY_TEAM_FAIL,
  TEAMS_CLEAN_UP,
  TEAMS_ADD_USER_FAIL,
  TEAMS_ADD_USER_INIT,
  TEAMS_ADD_USER_SUCCESS,
  TEAMS_REMOVE_USER_FAIL,
  TEAMS_REMOVE_USER_INIT,
  TEAMS_REMOVE_USER_SUCCESS,
  TEAMS_SELECTED_TEAM,
  TEAMS_FETCH_TEAM_USERS_FAIL,
  TEAMS_FETCH_TEAM_USERS_INIT,
  TEAMS_FETCH_TEAM_USERS_SUCCESS,
} from 'state/actions/teams';

const initialState = {
  teamsList: [],
  usersList: [],
  loadingTeams: false,
  loadingUsers: false,
  error: null,
  success: false,
  deleted: false,
  teamId: 0,
};

export const teamsReducer = createReducer(
  {
    [TEAMS_FETCH_TEAMS_INIT]: () => ({
      ...initialState,
      loadingTeams: true,
    }),
    [TEAMS_FETCH_TEAMS_SUCCESS]: (state, payload) => ({
      ...state,
      teamsList: payload.teams,
      loadingTeams: false,
      error: null,
    }),
    [TEAMS_FETCH_TEAMS_FAIL]: (state, payload) => ({
      ...state,
      loadingTeams: false,
      error: payload.error,
    }),
    [TEAMS_DELETE_TEAM_INIT]: (state) => ({
      ...state,
      loadingTeams: true,
    }),
    [TEAMS_DELETE_TEAM_SUCCESS]: (state, payload) => ({
      ...state,
      teamsList: state.teamsList.filter((elem) => elem.id !== payload.id),
      loadingTeams: false,
      error: null,
      deleted: true,
    }),
    [TEAMS_DELETE_TEAM_FAIL]: (state, payload) => ({
      ...state,
      loadingTeams: false,
      error: payload.error,
    }),
    [TEAMS_CREATE_TEAM_INIT]: (state) => ({
      ...state,
      loadingTeams: true,
    }),
    [TEAMS_CREATE_TEAM_SUCCESS]: (state, payload) => ({
      ...state,
      teamsList: state.teamsList.concat(payload.team),
      loadingTeams: false,
      error: null,
      success: true,
    }),
    [TEAMS_CREATE_TEAM_FAIL]: (state, payload) => ({
      ...state,
      loadingTeams: false,
      error: payload.error,
    }),
    [TEAMS_MODIFY_TEAM_INIT]: (state) => ({
      ...state,
      loadingTeams: true,
    }),
    [TEAMS_MODIFY_TEAM_SUCCESS]: (state, payload) => ({
      ...state,
      teamsList: !state.teamsList
        ? []
        : state.teamsList.map((elem) => {
            if (elem.id === payload.id) {
              return {
                ...elem,
                name: payload.team.name,
                description: payload.team.description,
                logoUrl: payload.team.logoUrl || null,
              };
            }
            return elem;
          }),
      loadingTeams: false,
      error: null,
      success: true,
    }),
    [TEAMS_MODIFY_TEAM_FAIL]: (state, payload) => ({
      ...state,
      loadingTeams: false,
      error: payload.error,
    }),
    [TEAMS_CLEAN_UP]: (state) => ({
      ...state,
      loadingTeams: false,
      error: null,
      success: false,
      deleted: false,
    }),
    [TEAMS_ADD_USER_FAIL]: (state, payload) => ({
      ...state,
      loadingUsers: false,
      error: payload.error,
    }),
    [TEAMS_ADD_USER_INIT]: (state) => ({
      ...state,
      loadingUsers: true,
    }),
    [TEAMS_ADD_USER_SUCCESS]: (state, payload) => ({
      ...state,
      loadingUsers: false,
      error: null,
      success: true,
      teamsList: state.teamsList.map((value) => {
        if (value.id === payload.teamId) {
          return { ...value, users: payload.users };
        }
        return value;
      }),
      usersList: [...state.usersList, payload.user],
    }),
    [TEAMS_REMOVE_USER_FAIL]: (state, payload) => ({
      ...state,
      loadingUsers: false,
      error: payload.error,
    }),
    [TEAMS_REMOVE_USER_INIT]: (state) => ({
      ...state,
      loadingUsers: true,
    }),
    [TEAMS_REMOVE_USER_SUCCESS]: (state, payload) => ({
      ...state,
      loadingUsers: false,
      error: null,
      teamsList: state.teamsList.map((value) => {
        if (value.id === payload.teamId) {
          return { ...value, users: payload.users };
        }
        return value;
      }),
      usersList: state.usersList.filter((value) => value.id !== payload.userId),
      deleted: true,
    }),
    [TEAMS_SELECTED_TEAM]: (state, payload) => ({
      ...state,
      teamId: payload.teamId,
    }),
    [TEAMS_FETCH_TEAM_USERS_INIT]: (state) => ({
      ...state,
      loadingUsers: true,
    }),
    [TEAMS_FETCH_TEAM_USERS_SUCCESS]: (state, payload) => ({
      ...state,
      usersList: payload.users || [],
      loadingUsers: false,
      error: null,
    }),
    [TEAMS_FETCH_TEAM_USERS_FAIL]: (state, payload) => ({
      ...state,
      loadingUsers: false,
      error: payload.error,
    }),
  },
  initialState
);
