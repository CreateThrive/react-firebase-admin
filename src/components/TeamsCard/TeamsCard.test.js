import { fireEvent } from '@testing-library/react';
import React from 'react';
import * as reactRedux from 'react-redux';

import * as teamsActions from 'state/actions/teams';

import TeamsCard from '.';

describe('<TeamsCard /> rendering', () => {
  const dispatchMock = jest.fn();

  const state = {
    auth: { userData: { isAdmin: true } },
    teams: {
      success: false,
      deleted: false,
      loadingUsers: false,
      loadingTeams: false,
      teamsList: [
        {
          id: 'teamTestId',
          name: 'testTeam',
          createdBy: 'test',
          createdAt: '11/12/2020',
        },
      ],
      usersList: [],
    },
    users: {
      data: [
        {
          email: 'test@gmail.com',
          name: 'testUser',
          location: 'Montevideo, Uruguay',
          isAdmin: true,
          file: null,
          id: 'testUser',
          logoUrl: 'some logoUrl',
          createdAt: '11/12/2020',
        },
      ],
    },
  };

  beforeEach(() => {
    jest
      .spyOn(reactRedux, 'useDispatch')
      .mockImplementation(() => dispatchMock);
    jest.spyOn(teamsActions, 'teamsCleanUp').mockImplementation(jest.fn);
  });
  it('should render without crashing', () => {
    const { component } = renderWithProviders(<TeamsCard />)({ ...state });

    expect(component.asFragment()).toMatchSnapshot();
  });

  it('should not render any buttons for non-admins', () => {
    const { component } = renderWithProviders(<TeamsCard />)({
      ...state,
      auth: {
        userData: {
          email: 'test@gmail.com',
          name: 'testUser',
          location: 'Montevideo, Uruguay',
          isAdmin: false,
          file: null,
          id: 'testUser',
          logoUrl: 'some logoUrl',
          createdAt: '11/12/2020',
        },
      },
    });

    expect(component.queryByRole('button')).toBeNull();
  });

  it('should render buttons for admins', () => {
    const { component } = renderWithProviders(<TeamsCard />)({
      ...state,
    });

    expect(component.getAllByRole('button')).toBeTruthy();
  });

  it('should open TeamModal when clicking add button', () => {
    const { component } = renderWithProviders(<TeamsCard />)({
      ...state,
    });

    fireEvent.click(component.getByText('Add Team'));

    expect(component.getByText('Team Information')).toBeTruthy();
  });
});
