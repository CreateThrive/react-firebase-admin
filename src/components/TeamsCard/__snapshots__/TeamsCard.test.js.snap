// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TeamsCard /> rendering should render without crashing 1`] = `
<DocumentFragment>
  <div
    class="card has-height-medium"
  >
    <header
      class="card-header"
    >
      <p
        class="card-header-title"
      >
        <span
          class="icon"
        >
          <i
            class="mdi mdi-account-group default"
          />
        </span>
        <span>
          Available teams
        </span>
      </p>
      <button
        class="button"
        type="button"
      >
        <span
          class="icon"
        >
          <i
            class="mdi mdi-account-multiple-plus default"
          />
        </span>
        <span>
          Add Team
        </span>
      </button>
    </header>
    <div>
      <div
        class="field-body"
        style="padding: 0.5rem 0.25rem;"
      >
        <div
          class="field"
        >
          <div
            class="control is-expanded"
          >
            <input
              class="input search"
              placeholder="Search by name"
              style="height: 38px;"
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          class="field"
        >
          <div
            class="control is-expanded"
          >
            <div
              style="min-width: 220px;"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="card-content ps ps--active-y"
      style="padding: 0px 1.5rem; overflow: auto; max-height: 700px;"
    >
      <div
        class="media-list"
      >
        <div
          class="media"
          data-testid="media"
          style="margin-top: auto; margin-bottom: auto; padding-bottom: 1rem; cursor: pointer;"
        >
          <figure
            class="media-left"
          >
            <p
              class="image is-64x64"
            >
              <img
                alt="Profile logo preview"
                class="is-rounded"
                src="team-default-log.png"
              />
            </p>
          </figure>
          <div
            class="media-content"
          >
            <div
              class="content"
            >
              <p
                class="media-meta"
              >
                <strong>
                  testTeam
                </strong>
                 
              </p>
              <p />
            </div>
          </div>
          <div
            class="media-right"
          >
            <div
              class="field is-grouped is-grouped-multiline"
            >
              <div
                class="control"
              >
                <span
                  class="icon has-text-info"
                  data-testid="modify"
                  style="cursor: pointer;"
                >
                  <i
                    class="mdi mdi-pencil"
                  />
                </span>
              </div>
              <div
                class="control"
              >
                <span
                  class="icon has-text-danger"
                  data-testid="delete"
                  style="cursor: pointer;"
                >
                  <i
                    class="mdi mdi-cancel"
                  />
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
